Bugzilla/Net/S3.pm

```diff
@@ -0,0 +1,192 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+#
+# This Source Code Form is "Incompatible With Secondary Licenses", as
+# defined by the Mozilla Public License, v. 2.0.
+
+package Bugzilla::Net::S3;
+
+use 5.10.1;
+use strict;
+use warnings;
+
+use Digest::HMAC_SHA1;
+use HTTP::Date;
+use List::Util qw(none);
+use MIME::Base64 qw(encode_base64);
+use Moo;
+use Types::Standard qw(Str);
+use URI::Escape     qw(uri_escape_utf8);
+use XML::Simple;
+
+use Bugzilla::Error;
+use Bugzilla::Util qw(trim);
+
+extends 'Bugzilla::Net::Storage';
+
+has client_id  => (is => 'ro', required => 1, isa => Str);
+has secret_key => (is => 'ro', required => 1, isa => Str);
+
+##################
+# Public methods #
+##################
+
+# Return type of net storage
+sub data_type { return 's3'; }
+
+###################
+# Private methods #
+###################
+
+sub _get_method_path {
+  my ($self, $action, $key) = @_;
+  my $method;
+
+  if (none {$action eq $_} qw(add head get delete)) {
+    ThrowCodeError('net_storage_invalid_action', {action => $action});
+  }
+
+  if ($action eq 'add') {
+    $method = 'PUT';
+  }
+  elsif ($action eq 'head') {
+    $method = 'HEAD';
+  }
+  elsif ($action eq 'get') {
+    $method = 'GET';
+  }
+  elsif ($action eq 'delete') {
+    $method = 'DELETE';
+  }
+
+  return ($method, $self->_uri($key));
+}
+
+# EU buckets must be accessed via their DNS name. This routine figures out if
+# a given bucket name can be safely used as a DNS name.
+sub _is_dns_bucket {
+  my ($self, $bucketname) = @_;
+
+  if (length $bucketname > 63) {
+    return 0;
+  }
+
+  if (length $bucketname < 3) {
+    return undef;
```

Comment:
Assuming this is supposed to be a truthy value, can we return `0` for consistency? If we are returning `undef` here for some other reason, let's leave a comment explaining why.